### PR TITLE
TM-4848 manage post access api

### PR DESCRIPTION
**Pull request by M00staff**
Dual Merge: 
- [API PR](https://github.com/MetaPhase-Consulting/State-TalentMAP-API/pull/838)

[Ticket](https://metaphase.atlassian.net/browse/TM-4848)

### CODE REVIEW

File: src/Components/ManagePostAccess/ManagePostAccess.jsx [hunk 1]
```diff
@@ -174,18 +150,16 @@ const ManagePostAccess = () => {
...
+      setSelectedRoles([{ code: 2010, description: 'HRO/MO' }]);
+    } else setSelectedRoles([]);
   };
-
-  const personFilterToggle = () => {
-    // nothing happening for this yet. UI Only.
-    if (selectedPersons.length > 0) {
-      setSelectedPersons([]);
-    }
+  // TODO - How should these work?
```
Comment (M00staff) [inline src/Components/ManagePostAccess/ManagePostAccess.jsx:158]: all thats left to do is test on dev also need clarification on these checkboxes

File: src/Components/ManagePostAccess/ManagePostAccess.jsx [hunk 2]
```diff
@@ -295,30 +215,31 @@ const ManagePostAccess = () => {
...
-                labelKey="name"
-                disabled={filtersLoading}
+                labelKey="description"
+                disabled={managePostFiltersIsLoading}
               />
             </div>
             <div className="filter-div">
               <div className="label">Position:</div>
-              <div className="post-access-container-cb">
+              {/* DOUBLE CHECK THIS IS SUPPOSED TO BE ROLE NOT POSITION */}
```
Comment (M00staff) [inline src/Components/ManagePostAccess/ManagePostAccess.jsx:224]: spreadsheet has this as Role, design has it as position

File: src/Components/ManagePostAccess/ManagePostAccess.jsx [hunk 3]
```diff
@@ -38,30 +34,27 @@ const ManagePostAccess = () => {
...
   // Filter Options
-  const genericFilters$ = genericFilters?.filters || [];
-  const positionOptions = getGenericFilterOptions(genericFilters$, 'region', 'short_description');
-  const roleOptions = getGenericFilterOptions(genericFilters$, 'skill', 'code');
-  const { data: orgs, loading: orgsLoading } = useDataLoader(api().get, '/fsbid/agenda_employees/reference/current-organizations/');
-  const organizationOptions = (orgs?.data?.length && nameSort(orgs?.data, 'name')) || [];
+  const peopleOptionsHRO = managePostFilters?.personFilters
+    ?.filter(person => person.skillCode === '2010') || [];
+  const positionOptionsHRO = managePostFilters?.positionFilters
+    ?.filter(position => position.skillCode === '2201') || [];
```
Comment (M00staff) [inline src/Components/ManagePostAccess/ManagePostAccess.jsx:44]: i still need to confirm these are the correct skill codes for HBO/MO only persons and positions 

Tim said he'd ask Robin

Comment (M00staff) [reply src/Components/ManagePostAccess/ManagePostAccess.jsx:44]: Filtering the filters on the frontend, by their skill code - apparently that's how this checkbox is supposed to work .. though it's wonky in FSBID  🤷

image

File: src/Components/ManagePostAccess/ManagePostAccess.jsx [hunk 4]
```diff
@@ -273,175 +131,184 @@ const ManagePostAccess = () => {
...
+                    />
+                    <Picky
+                      {...pickyProps}
+                      placeholder="Select Position(s)"
+                      value={selectedPositions}
+                      options={positionOptions}
+                      onChange={setSelectedPositions}
+                      valueKey="code"
+                      labelKey="description"
+                      disabled={managePostFiltersIsLoading || personSelected}
```
Comment (M00staff) [inline src/Components/ManagePostAccess/ManagePostAccess.jsx:171]: disable Person or Position if one or the other is selected - its a requirement

File: src/Components/ManagePostAccess/ManagePostAccess.jsx [hunk 5]
```diff
@@ -273,175 +131,184 @@ const ManagePostAccess = () => {
...
+                  </div>
+                </div>
+                <div className="filter-div">
+                  <div className="label">Person:</div>
+                  <div className="post-access-container-cb">
+                    <CheckBox
+                      label="HRO/MO Only"
+                      small
+                      value={personHRO}
+                      onCheckBoxClick={() => setPersonHRO(!personHRO)}
```
Comment (M00staff) [inline src/Components/ManagePostAccess/ManagePostAccess.jsx:182]: The HRO/MO filter, filters on the frontend Positions or Persons with a specific skill code.

File: src/sass/_postAccess.scss [hunk 6]
```diff
@@ -65,3 +65,30 @@
...
+  width: 25%;
+  border: 2px solid $tm-gray;
+  margin: 10px;
+  overflow-y: scroll;
+  box-shadow: $tm-gray 1px;
+}
+
+.mpa-grant-box-item {
+  display: grid;
+  grid-template-columns: repeat(auto-fit, minmax(85%, max-content));
```
Comment (M00staff) [inline src/sass/_postAccess.scss:86]: This was kind of a pain to get working correctly - and still isn't perfect - really long names throw it off

image

File: src/Components/ManagePostAccess/ManagePostAccess.jsx [hunk 7]
```diff
@@ -2,190 +2,104 @@ import { useEffect, useState } from 'react';
...
-    'post-access-orgs': selectedOrgs.map(orgObject => (orgObject?.code)),
-    'post-access-persons': selectedPersons.map(gradeObject => (gradeObject?.code)),
-    'post-access-roles': selectedRoles.map(skillObject => (skillObject?.code)),
-    page,
-    limit,
-  });
+  const peopleOptionsHRO = managePostFilters?.personFilters?.filter(
+    person => person.skillCode === '2010' || person.skillCode === '2201') || [];
+  const positionOptionsHRO = managePostFilters?.positionFilters?.filter(
+    position => position.skillCode === '2201' || position.skillCode === '2010') || [];
```
Comment (M00staff) [inline src/Components/ManagePostAccess/ManagePostAccess.jsx:41]: The HRO/MO filter, filters on the frontend Positions or Persons with a specific skill code.

File: src/Components/ManagePostAccess/ManagePostAccess.jsx [hunk 8]
```diff
@@ -273,179 +133,184 @@ const ManagePostAccess = () => {
...
-                    </table>
+              ))}
+            </div>
 
+            <div className="mpa-grant-box">
+              <div>Positions:</div>
+              {personSelected && (
+                <div className="mpa-grant-box-item">
+                  Grant Access to either Persons or Positions.
+                </div>
```
Comment (M00staff) [inline src/Components/ManagePostAccess/ManagePostAccess.jsx:247]: notice pops up when Person or Position filters are selected - noting only one can be passed in

File: src/Components/ManagePostAccess/ManagePostAccess.jsx [hunk 9]
```diff
@@ -2,190 +2,104 @@ import { useEffect, useState } from 'react';
...
+    useState(managePostSelections?.selectedPersons || []);
   const [clearFilters, setClearFilters] = useState(false);
-  const [selectAll, setSelectAll] = useState(false);
+  const [personHRO, setPersonHRO] = useState(false);
+  const [positionHRO, setPositionHRO] = useState(false);
+
+  const personSelected = selectedPersons.length > 0;
+  const positionSelected = selectedPositions.length > 0;
+  const requirementsMet =
+    (personSelected || positionSelected) && selectedOrgs.length > 0 && selectedRoles.length > 0;
```
Comment (M00staff) [inline src/Components/ManagePostAccess/ManagePostAccess.jsx:35]: only show the 'Grant Access' button when these conditions are met

File: src/Components/ManagePostAccess/ManagePostAccess.jsx [hunk 10]
```diff
@@ -273,179 +133,184 @@ const ManagePostAccess = () => {
...
+                </div>
+              ))}
             </div>
+
+          </div>
+
+        </>
       }
 
+      { requirementsMet && (
```
Comment (M00staff) [inline src/Components/ManagePostAccess/ManagePostAccess.jsx:308]: only show the 'Grant Access' button when the right conditions are met 

Person or Position selected + at least one Role & Org

File: src/actions/managePostAccess.js [hunk 11]
```diff
@@ -190,25 +74,57 @@ export function managePostFiltersFetchDataErrored(bool) {
...
-export function managePostFiltersFetchData() {
+export function managePostFetchFilters() {
   return (dispatch) => {
     batch(() => {
-      dispatch(managePostFiltersFetchDataSuccess({}));
-      dispatch(managePostFiltersFetchDataLoading(false));
+      dispatch(managePostFetchFiltersLoading(true));
+      dispatch(managePostFetchFiltersErrored(false));
     });
+    const endpoint = 'fsbid/post_access/filters/';
```
Comment (M00staff) [inline src/actions/managePostAccess.js:105]: same endpoint as the other page - but different actions, so we can fire different toasts 🤷

File: src/reducers/managePostAccess/managePostAccess.js [hunk 12]
```diff
@@ -25,32 +25,6 @@ export function managePostEdit(state = {}, action) {
...
-}
-
-export function managePost(state = {}, action) {
-  switch (action.type) {
-    case 'MANAGE_POST_FETCH_SUCCESS':
-      return action.results;
-    default:
-      return state;
-  }
-}
```
Comment (M00staff) [inline src/reducers/managePostAccess/managePostAccess.js:53]: since we're not rendering any data outside of the filter - these were not needed

**Comment by elizabeth-jimenez** [comment]
Starting my trickle review - will let you know when all feedback has been given

**Comment by elizabeth-jimenez** [comment]
image

Wondering if we can merge the old designs with new UI by counting the People or Positions being granted roles.
`5 People Selected - Grant Access`
`5 Positions Selected - Grant Access`

File: src/sass/_postAccess.scss (hunk 6 above)
Comment (elizabeth-jimenez) [reply src/sass/_postAccess.scss:86]: image
try this out - seems to work

File: src/Components/ManagePostAccess/ManagePostAccess.jsx [hunk 13]
```diff
@@ -2,190 +2,104 @@ import { useEffect, useState } from 'react';
...
-   useState(managePostSelections?.selectedCountries || []);
   const [selectedPositions, setSelectedPositions] =
    useState(managePostSelections?.selectedPositions || []);
   const [selectedOrgs, setSelectedOrgs] = useState(managePostSelections?.selectedOrgs || []);
-  const [selectedPersons, setSelectedPersons] = useState(managePostSelections?.selectedGrade || []);
-  const [selectedRoles, setSelectedRoles] = useState(managePostSelections?.selectedSkills || []);
-  const [checkedPostIds, setCheckedPostIds] = useState([]);
+  const [selectedRoles, setSelectedRoles] = useState(managePostSelections?.selectedRoles || []);
+  const [selectedPersons, setSelectedPersons] =
+    useState(managePostSelections?.selectedPersons || []);
```
Comment (elizabeth-jimenez) [inline src/Components/ManagePostAccess/ManagePostAccess.jsx:27]: The dropdowns on this page are being used more as form inputs than as filters, so we should probably not cache their inputs and do away with all of the `managePostSelections` stuff

**Comment by elizabeth-jimenez** [comment]
keeping the inputs on error makes sense, but on success it probably makes more sense to clear them out

**Comment by elizabeth-jimenez** [comment]
We've gotten into the habit of setting up success, loading, and error redux states for our Edit, but it's cleaner to just use the same Read redux state ones, since we always refresh on an Edit to show users the latest data.

Edit Action Success -> toast in Edit Action -> call Read Action -> Read takes over

That being said, this page deviates from that pattern a bit, bc of my comment here https://github.com/MetaPhase-Consulting/State-TalentMAP/pull/2687#issuecomment-1743482901 where we want to listen to the success to clear the user inputs (and there are no 'results' for us to call)

so while keeping `managePostEditSuccess` Action makes sense (to clear user inputs), we can go ahead and remove `managePostEditErrored` and `managePostEditLoading`

File: src/Components/ManagePostAccess/ManagePostAccess.jsx [hunk 14]
```diff
@@ -2,190 +2,104 @@ import { useEffect, useState } from 'react';
...
 
-  // Handle Pagination
-  useEffect(() => {
-    setCheckedPostIds([]);
-    setSelectAll(false);
-    fetchAndSet(false);
-  }, [page]);
-
 
   const resetFilters = () => {
```
Comment (elizabeth-jimenez) [inline src/Components/ManagePostAccess/ManagePostAccess.jsx:88]: we'll want to include the HRO/MO checkboxes

File: src/Components/ManagePostAccess/ManagePostAccess.jsx [hunk 15]
```diff
@@ -2,190 +2,104 @@ import { useEffect, useState } from 'react';
...
     setSelectedPositions([]);
     setSelectedOrgs([]);
     setSelectedPersons([]);
     setSelectedRoles([]);
     setClearFilters(false);
   };
 
-  const noResults = managePostData?.results?.length === 0;
-  const getOverlay = () => {
-    let overlay;
```
Comment (elizabeth-jimenez) [inline src/Components/ManagePostAccess/ManagePostAccess.jsx:139]: 🙌

File: src/Components/ManagePostAccess/ManagePostAccess.jsx [hunk 16]
```diff
@@ -273,179 +133,184 @@ const ManagePostAccess = () => {
...
+                </div>
+              </>
+            }
+
           </div>
         </div>
       </div>
 
 
       {
```
Comment (elizabeth-jimenez) [inline src/Components/ManagePostAccess/ManagePostAccess.jsx:220]: L220 {<></>} wrapping fragment not needed

File: src/Components/ManagePostAccess/ManagePostAccess.jsx (hunk 7 above)
Comment (elizabeth-jimenez) [reply src/Components/ManagePostAccess/ManagePostAccess.jsx:41]: nice job handling this

File: src/Components/ManagePostAccess/ManagePostAccess.jsx [hunk 17]
```diff
@@ -2,190 +2,104 @@ import { useEffect, useState } from 'react';
...
-    }
-    if (resetPage) {
-      setPage(1);
-    }
+  const filters = [
+    selectedPositions,
+    selectedOrgs,
+    selectedPersons,
+    selectedRoles,
+  ];
```
Comment (elizabeth-jimenez) [inline src/Components/ManagePostAccess/ManagePostAccess.jsx:61]:
```suggestion
  const filters = [
    selectedPositions,
    selectedOrgs,
    selectedPersons,
    selectedRoles,
    ...personHRO ? ['Y'] : [],
    ...positionHRO ? ['Y'] : [],
  ];
```
to capture the checkboxes

File: src/Components/ManagePostAccess/ManagePostAccess.jsx [hunk 18]
```diff
@@ -2,190 +2,104 @@ import { useEffect, useState } from 'react';
...
 
   // Initial Render
   useEffect(() => {
     dispatch(saveManagePostSelections(getCurrentInputs()));
-    dispatch(filtersFetchData(genericFilters));
+    dispatch(managePostFetchFilters());
   }, []);
 
   // Re-Render on Filter Selections
   useEffect(() => {
```
Comment (elizabeth-jimenez) [inline src/Components/ManagePostAccess/ManagePostAccess.jsx:77]: we can get rid of

File: src/actions/managePostAccess.js [hunk 19]
```diff
@@ -75,103 +28,34 @@ export function managePostEditSuccess(results) {
   };
 }
 
-let cancel;
 export function managePostEdit(positions) {
```
Comment (elizabeth-jimenez) [inline src/actions/managePostAccess.js:31]: rename param to something more descriptive

File: src/Components/ManagePostAccess/ManagePostAccess.jsx [hunk 20]
```diff
@@ -273,179 +133,184 @@ const ManagePostAccess = () => {
           </div>
 
           <div className="usa-width-one-whole position-search--filters--pv-man results-dropdown">
```
Comment (elizabeth-jimenez) [inline src/Components/ManagePostAccess/ManagePostAccess.jsx:135]: we'll want to set up a unique one for manage post access `position-search--filters--pv-man`

File: src/reducers/managePostAccess/index.js [hunk 21]
```diff
@@ -1,25 +1,19 @@
 import {
-  managePost,
   managePostEdit,
   managePostEditErrored,
   managePostEditLoading,
```
Comment (elizabeth-jimenez) [inline src/reducers/managePostAccess/index.js:4]: already commented on another place, but can remove these two

File: src/sass/_postAccess.scss [hunk 22]
```diff
@@ -65,3 +65,30 @@
...
+
+.mpa-grant-box-wrapper {
+  display: flex;
+  align-content: center;
+  flex-direction: row;
+}
+
+.mpa-grant-box {
+  height: 500px;
+  width: 25%;
```
Comment (elizabeth-jimenez) [inline src/sass/_postAccess.scss:77]: nice job on setting this up

File: src/sass/_postAccess.scss [hunk 23]
```diff
@@ -65,3 +65,30 @@
   position: relative;
```
Comment (elizabeth-jimenez) [inline src/sass/_postAccess.scss:65]: since we reworked the UI, there is a lot of css we can delete now, like `custom-table`

Comment (elizabeth-jimenez) [reply src/sass/_postAccess.scss:65]: First 63 lines look like they can go

**Comment by elizabeth-jimenez** [comment]
Actions file seems to have a duplicate for filters error state `managePostFiltersFetchDataErrored`

**Comment by elizabeth-jimenez** [comment]
Looks like action file can be cleaned up to just include the following 🤔 
image

File: src/Components/ManagePostAccess/ManagePostAccess.jsx [hunk 24]
```diff
@@ -273,179 +133,184 @@ const ManagePostAccess = () => {
...
+              :
+              <>
+                <div className="filter-div">
+                  <div className="label">Organization:</div>
+                  <Picky
+                    {...pickyProps}
+                    placeholder="Select Organization(s)"
+                    value={selectedOrgs}
+                    options={organizationOptions}
+                    onChange={setSelectedOrgs}
```
Comment (elizabeth-jimenez) [inline src/Components/ManagePostAccess/ManagePostAccess.jsx:149]: I need to verify how this is being handled on the BE - filters vs inputs

**Comment by elizabeth-jimenez** [comment]
Trickle Review Complete

File: src/Components/ManagePostAccess/ManagePostAccess.jsx [hunk 25]
```diff
@@ -2,190 +2,104 @@ import { useEffect, useState } from 'react';
...
-      overlay = <Alert type="error" title="Error loading results" messages={[{ body: 'Please try again.' }]} />;
-    } else if (noResults) {
-      overlay = <Alert type="info" title="No results found" messages={[{ body: 'Please broaden your search criteria and try again.' }]} />;
-    } else {
-      return false;
-    }
-    return overlay;
-  };
-
   const submitGrantAccess = () => {
```
Comment (elizabeth-jimenez) [inline src/Components/ManagePostAccess/ManagePostAccess.jsx:96]: https://github.com/MetaPhase-Consulting/State-TalentMAP-API/pull/838/files#r1343093597 this is where we want to filter just the codes for the BE

File: src/sass/_postAccess.scss (hunk 6 above)
Comment (M00staff) [reply src/sass/_postAccess.scss:86]: NICE!

File: src/sass/_postAccess.scss (hunk 23 above)
Comment (M00staff) [reply src/sass/_postAccess.scss:65]: WOW - can't believe i spaced on that!

File: src/Components/ManagePostAccess/ManagePostAccess.jsx (hunk 18 above)
Comment (M00staff) [reply src/Components/ManagePostAccess/ManagePostAccess.jsx:77]: I'm not sure we can?

The re-render is informing the clear filters button ... 🤔

File: src/Components/ManagePostAccess/ManagePostAccess.jsx (hunk 25 above)
Comment (M00staff) [reply src/Components/ManagePostAccess/ManagePostAccess.jsx:96]: so, is this OK as is - i see another comment below the one you linked to?

**Comment by M00staff** [comment]
i think i got everything, outside of those 2 questions

File: src/actions/managePostAccess.js [hunk 26]
```diff
@@ -9,206 +8,87 @@ import {
...
 }
 
-let cancel;
-export function managePostEdit(positions) {
+export function managePostEdit(accessors) {
   return (dispatch) => {
-    if (cancel) { cancel('cancel'); }
-    dispatch(managePostEditLoading(true));
-    dispatch(managePostEditErrored(false));
     api()
```
Comment (elizabeth-jimenez) [inline src/actions/managePostAccess.js:21]:
```
  return (dispatch) => {
    dispatch(managePostEditSuccess(false));
    api()
```
you'll want to reset Success state

File: src/Components/ManagePostAccess/ManagePostAccess.jsx (hunk 24 above)
Comment (elizabeth-jimenez) [reply src/Components/ManagePostAccess/ManagePostAccess.jsx:156]: verified - all gucci

File: src/Components/ManagePostAccess/ManagePostAccess.jsx (hunk 25 above)
Comment (elizabeth-jimenez) [reply src/Components/ManagePostAccess/ManagePostAccess.jsx:97]: Yea this is fine!

File: src/Components/ManagePostAccess/ManagePostAccess.jsx (hunk 18 above)
Comment (elizabeth-jimenez) [reply src/Components/ManagePostAccess/ManagePostAccess.jsx:67]: resolved in Slack

File: src/actions/managePostAccess.js (hunk 26 above)
Comment (elizabeth-jimenez) [reply src/actions/managePostAccess.js]: i pushed up the fix

**Review by elizabeth-jimenez — APPROVED** [review]
Very nice

**Pull request merged**